### PR TITLE
Feature/0293 oem exclusive app support

### LINKED ISSUE #1588: [SDL 0293] Enable OEM exclusive apps support
**Proposal:** [Enable OEM exclusive apps support](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0293-vehicle-type-filter.md)

> This feature will enable SDL adopters to provide exclusive apps to their users depending on vehicle type. The proposal describes a way to share vehicle type information before sending the Register App Interface request.

**Review:** https://github.com/smartdevicelink/sdl_evolution/issues/971

**Steering Committee Decision:** 
The Steering Committee voted to accept this proposal with the following revisions:

The author is to update the proposal to include the requested requirement for the Java Suite Library to have two pull requests submitted for this feature. The reasoning as to why is documented in [this comment ](https://github.com/smartdevicelink/sdl_evolution/issues/971#issuecomment-730557977)and should be included in the proposal.

The proposal .md file was updated to reflect these revisions on 12/14/2020.

### PR DESCRIPTION
**Pull request by kostyaBoss**
Fixes #1588

This PR is **[ready]** for review.

### Risk
This PR makes **[minor]** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [x] I have tested Android, Java SE, and Java EE

#### Core Tests
[List of tests performed against Core and behaviors verified]

[Core version / branch / commit hash / module tested against](https://github.com/LuxoftSDL/sdl_core/tree/feature/sdl_0293_enable_oem_exclusive_apps_support)
[HMI name / version / branch / commit hash / module tested against](https://github.com/LuxoftSDL/sdl_hmi/tree/feature/sdl_0293_enable_oem_exclusive_apps_support)

### Summary
Changes according to #1588

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: base/src/main/java/com/smartdevicelink/managers/BaseSdlManager.java
Comment (yoooriii) [inline base/src/main/java/com/smartdevicelink/managers/BaseSdlManager.java:893]: maybe we had better put it other way:
`return vehicleMakes.contains(type);`

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlReceiver.java
Comment (ashwink11) [inline android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlReceiver.java:47]: Please return true from this method. Since this is a sample app, it should connect to all vehicles.

File: android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java
Comment (ashwink11) [inline android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java:183]: This section will read vehicle type info if available and close BT socket. Please check [here](https://raw.githubusercontent.com/smartdevicelink/sdl_evolution/master/assets/proposals/0293-vehicle-type-filter/OEM_APPS_2.png).

File: base/src/main/java/com/smartdevicelink/transport/TransportConstants.java
Comment (ashwink11) [inline base/src/main/java/com/smartdevicelink/transport/TransportConstants.java:93]: The constant define in proposal is as below. Please change it.
`public static final String VEHICLE_INFO = "vehicle_info";`

**Review by ashwink11 — CHANGES_REQUESTED** [review]

File: android/hello_sdl_android/src/main/res/layout/activity_main.xml
Comment (ashwink11) [inline android/hello_sdl_android/src/main/res/layout/activity_main.xml:16]: Could you please let us know why buttons were added in app?

**Comment by a-prykho** [comment]
@santhanamk could you please start Ford review? Thanks

File: base/src/main/java/com/smartdevicelink/session/ISdlSessionListener.java
Comment (santhanamk) [inline base/src/main/java/com/smartdevicelink/session/ISdlSessionListener.java:87]: @kostyaBoss On line 87 does this method `onVehicleTypeReceived` need JavaDoc comments around it like the other methods?

Comment (santhanamk) [inline base/src/main/java/com/smartdevicelink/session/ISdlSessionListener.java:87]: @kostyaBoss Please update the file `MockInterfaceBroker.java` to include `onVehicleTypeReceived(...)`.  It doesn't build.

File: base/src/main/java/com/smartdevicelink/transport/TransportConstants.java
Comment (santhanamk) [inline base/src/main/java/com/smartdevicelink/transport/TransportConstants.java:93]: @kostyaBoss You can remove the deleted lines here around line 92.

File: android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java
Comment (santhanamk) [inline android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java:180]: @kostyaBoss For lines 180-186 are these fields `model`, `modelYear`, `trim` optional?  Do you need to check if these are `null` as well?

File: android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java
Comment (santhanamk) [inline android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java:61]: @kostyaBoss On line 61 you have removed the 2 argument constructor.  You may need to bring it back for backward compatibility.

File: android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java
Comment (santhanamk) [inline android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java:73]: @kostyaBoss `SdlAppInfoTests.java`  is not building because it is referring to the 2 argument constructor.  Please correct this file so that it builds.  Bringing back the 2 argument constructor may help resolve it.

**Review by santhanamk — COMMENTED** [review]
@kostyaBoss I have reviewed this PR, and have given some suggestions.  Please take a look whenever you get the chance.

File: android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java:221]: I don't think so, the SdlSessionTests are testing only the ServiceListeners

File: base/src/main/java/com/smartdevicelink/managers/lifecycle/BaseLifecycleManager.java
Comment (kostyaBoss) [reply base/src/main/java/com/smartdevicelink/managers/lifecycle/BaseLifecycleManager.java:940]: @santanamk, we don't need to know about the manager, because the particular one will be called using delegated instances, hence the reference is redundant as for me

File: base/src/main/java/com/smartdevicelink/session/ISdlSessionListener.java
Comment (kostyaBoss) [reply base/src/main/java/com/smartdevicelink/session/ISdlSessionListener.java:87]: @santanamk, I don't have the valid one, that's why I didn't add anything

File: android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java:61]: I don't think we need to keep the backward compatibility in this way because in this case the all new VehicleType information would be lost, because of null-context issue

Comment (santhanamk) [reply android/sdl_android/src/main/java/com/smartdevicelink/util/SdlAppInfo.java:61]: @kostyaBoss `SdlAppInfoTests.java` doesn't build because it is using the 2 argument constructor.  Can you fix the build issues in `SdlAppInfoTests.java`?

File: base/src/main/java/com/smartdevicelink/transport/TransportConstants.java
Comment (santhanamk) [inline base/src/main/java/com/smartdevicelink/transport/TransportConstants.java:92]: @kostyaBoss Please place the line `public static final String PING_ROUTER_SERVICE_EXTRA = "ping.router.service";` right below line 91.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:70]: @kostyaBoss On line 70 I would use actual values for `VEHICLE_MAKE`.  You could use `Ford` instead of `sdl`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:71]: @kostyaBoss On line 71 I would use actual values for `VEHICLE_MODEL`.  You could use `Mustang` instead of `sdl`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:72]: @kostyaBoss On line 72 I would use actual values for `VEHICLE_TRIM`.  You could use `GT` instead of `sdl`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:85]: @kostyaBoss On line 85 you can replace `sdl` with `Ford`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:86]: @kostyaBoss On line 86 you can replace `sdl` with `Mustang`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:87]: @kostyaBoss On line 87 you can replace `sdl` with `GT`.

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java
Comment (santhanamk) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/protocol/SdlPacketTests.java:82]: @kostyaBoss I would swap lines 82 and 83.  (So the order ends up being `VEHICLE_MAKE`, `VEHICLE_MODEL`, `VEHICLE_TRIM`, `VEHICLE_MODEL_YEAR`) like the other two sections. (70-73 for example)

File: android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java
Comment (santhanamk) [inline android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java:290]: @kostyaBoss This method `getVehicleTypeFromPrefs()` on line 290 is not being used anywhere.  Do you need it or can it be deleted?

**Review by santhanamk — COMMENTED** [review]
@kostyaBoss I have reviewed your latest commits, and have provided some feedback.  Please take a look whenever you get the chance.

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java]: @santanamk the thing is - the context is needed in the base class, if we will leave it as private - there won't be variable access

File: android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/transport/utl/SdlDeviceListener.java]: @santhanamk just mirroring the second PR, let's keep it

**Comment by kostyaBoss** [comment]
@santhanamk I've processed the comments, please, re-review

**Review by santhanamk — COMMENTED** [review]
@kostyaBoss I have reviewed the PR.  I have one suggestion.  Please take a look whenever you get the chance.

**Comment by kostyaBoss** [comment]
@santhanamk Covered as much as possible, added a bunch of tests to cover new functions. There is also a failing test after merging the latest develop branch, related to the RemoteDisplayTests. Please, re-review, let me know if there is anything I should adjust

**Review by santhanamk — APPROVED** [review]
@kostyaBoss Code looks good.  The `SdlRemoteDisplayTest` passed here as well.  I have approved this PR.

**Comment by jordynmackool** [comment]
@kostyaBoss and @santhanamk is this PR ready for Livio review?

**Comment by a-prykho** [comment]
Hi @jordynmackool ,this PR is ready for Livio review

**Comment by jordynmackool** [comment]
@kostyaBoss @a-prykho @santhanamk 

There are still unresolved comments on this PR. Can you please have the Ford reviewer either resolve the comments if they are complete or have Luxoft make the changes requested.  

Once all the comments are resolved please ensure testing for this PR was completed before re-tagging us for review.

**Comment by santhanamk** [comment]
> @kostyaBoss @a-prykho @santhanamk
> 
> There are still unresolved comments on this PR. Can you please have the Ford reviewer either resolve the comments if they are complete or have Luxoft make the changes requested.
> 
> Once all the comments are resolved please ensure testing for this PR was completed before re-tagging us for review.

Hi @jordynmackool this PR is ready for Livio review.  I am not able to resolve the comments for some reason, but all the comments were addressed.

File: android/hello_sdl_android/src/main/res/layout/activity_main.xml
Comment (santhanamk) [reply android/hello_sdl_android/src/main/res/layout/activity_main.xml:18]: Hi @kostyaBoss Do you need these buttons or can they be removed?

**Review by santhanamk — COMMENTED** [review]
@kostyaBoss Do you need the buttons in the app?

**Comment by kostyaBoss** [comment]
@santhanamk removed, re-review please

**Review by santhanamk — APPROVED** [review]
@kostyaBoss This looks good.  I am approving this PR.

**Comment by santhanamk** [comment]
Hi @jordynmackool this PR is ready for Livio review.

**Comment by jordynmackool** [comment]
@kostyaBoss the proposal markdown file has been updated per the revisions included in the accepted with revisions review issue: [SDL 0293 Revisions - Enable OEM exclusive apps support](https://github.com/smartdevicelink/sdl_evolution/issues/1108). 

Please make the needed updates to this PR and then tag @JulianKast to review when ready. Thanks!

**Comment by kostyaBoss** [comment]
@jordynmackool I've updated the pr according to the latest revision

**Review by JulianKast — CHANGES_REQUESTED** [review]
Hello, @kostyaBoss.

 Requested actions and changes: 

- Add a method to `SdlManagerListener.java` like the proposal states to let app developers handle checking `SystemInfo`. 
- `SystemInfo` needs to be added to our base folder `base/src/main/java/com/smartdevicelink/session/`.   
- Add `onSystemInfoReceived` to the implementation of `SdlManagerListener` in `SdlService` for both `hello_sdl_android` as well as `hello_sdl_java`. 
- Implement requested changes, clean up imports, fix unit tests related to requested changes and test with our JavaSE project in addition to our Android project

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java:72]: @JulianKast can't process this comment, otherwise, there will be no context access in BaseManager class which requires it to get vehicles from Manifest

Comment (joeygrover) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java:72]: @kostyaBoss you do not need the context. The implementation is incorrect according to the proposal. You should also **never** access a variable stored in a child class from a parent class without even checking if the instance in question is actually of the child class. The code that accesses this variable causes an instance exception in JavaSE which should have been caught if this was actually tested against JavaSE as you have stated. Please follow all of these requests as provided. Thank you.

Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java:72]: @joeygrover Thanks for your reply. I am confused, because in the BaseSdlManager we should process the systemInfo and manifest specific things. This is not possible without having the context reference (the screenshot is attached). On the other hand I didn't find any place where I can get the context from. Not sure If I can move context out of there
изображение

Comment (joeygrover) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java:72]: That is incorrect. The proposal states the developer is responsible for this exact logic. If you follow the review from @JulianKast you will see that the code here does not follow the proposal and creates disparity between the other platforms. The system info is handled by the developer through the listener they pass to the SdlManager. Please read through the entire review to understand where the issues are and address them as a whole. Thank you.

File: android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java:151]: @JulianKast Am I correct you are suggesting to remove this caching logic? According to the proposal, we should cache the vehicle type in case we've retrieved it in the RAI as well. Do we still need to remove it?

File: base/src/main/java/com/smartdevicelink/managers/lifecycle/BaseLifecycleManager.java
Comment (kostyaBoss) [reply base/src/main/java/com/smartdevicelink/managers/lifecycle/BaseLifecycleManager.java:942]: @JulianKast but there are legacy versions of devices (as far as I can understand) that are not going to send this in the StartServiceACK, that was the reason for marking this param as nullable. Should we add null check and form didReceiveSystemInfo depending on that?

Comment (kostyaBoss) [reply base/src/main/java/com/smartdevicelink/managers/lifecycle/BaseLifecycleManager.java:942]: @JulianKast after a little more deep investigation I've got your point on this, this comment is resolved

File: android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java
Comment (kostyaBoss) [reply android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java:61]: Thanks for your feedback, referencing to this [comment](https://github.com/smartdevicelink/sdl_java_suite/pull/1592#discussion_r568960468) this code is used to determine if it's RAI case currently and we should cache data locally which was posted on the proposal's explanation diagrams. Removing this part will cause removing all the caching on the RAI case, do we still need to remove it?

File: base/src/main/java/com/smartdevicelink/protocol/SdlProtocolBase.java
Comment (kostyaBoss) [reply base/src/main/java/com/smartdevicelink/protocol/SdlProtocolBase.java:286]: @JulianKast Thanks for your feedback, referencing to this [comment](https://github.com/smartdevicelink/sdl_java_suite/pull/1592#discussion_r568960468) this code is used to determine if it's RAI case currently and we should cache data locally which was posted on the proposal's explanation diagrams. Removing this part will cause removing all the caching on the RAI case, do we still need to remove it?

File: android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java
Comment (joeygrover) [reply android/sdl_android/src/main/java/com/smartdevicelink/session/SdlSession.java:151]: The information is only cached for the Android implementation when used with the router service start up functionality. Therefore the code here, does not belong in this PR. It should only be part of the PR that includes the specific Android changes.

**Comment by joeygrover** [comment]
Closing in favor of #1618. This pull request had a good amount of unnecessary changes, didn't compile for 2/3 of the libraries based on this project, and was taking a long time to get addressed. Due to the Sdl Core project and supporting projects coming to a development cycle end this week it was determined that the project maintainer could no longer wait and assume this PR was going to be in a state that was mergeable or testable. The PM has created a new PR that only contains the necessary changes so that we can test the Core side of the feature and ensure its inclusion in this release. The author of this PR is still responsible for the Android specific PR of the feature and that PR should be updated once #1618 is reviewed and merged. The authors of the other PRs related to this feature for the iOS and JavaScript Suite projects should use the new PR as a base of how to implement the feature. Thank you for your understanding.

**Pull request closed without merging**
